### ISSUE TITLE
Server Sent Events

### ISSUE THREAD

**Issue by tomchristie** [encode/httpx#1278]
It'd be really neat to offer built-in support for server sent events, which are a way for servers to stream events to an HTTP client.

(Similar to WebSockets, but uni-directional)

```python
with httpx.stream("GET", "https://www.example.org/sse") as response:
    for sse in response.iter_sse():
        print(sse.data, sse.event, sse.id, sse.retry)
```

We'd probably also want to include an `sse.json()` helper method.

Useful resources...

* https://www.html5rocks.com/en/tutorials/eventsource/basics/
* https://sysid.github.io/sse/
* https://pgjones.gitlab.io/quart/tutorials/broadcast_tutorial.html

Not sure if this has come up before or not, but it'd be a *really* neat feature for us to offer, for relatively little effort.

**Comment by simonw** [comment]
I really like this. SSE are a under-used feature of the web in my opinion - I could absolutely see myself hooking up a really simple message bus between two machines using SSE served by Starlette and received by HTTPX.

**Comment by florimondmanca** [comment]
FWIW, I had played with SSE myself in Bocadillo at the time, so just linking to the [user docs](https://bocadilloproject.github.io/guide/sse.html#sse-basics) and the [server-side encoding implementation](https://github.com/bocadilloproject/bocadillo/blob/master/bocadillo/sse.py) in case they could be useful here, as some extra reference material.

**Comment by havardthom** [comment]
+1 on this feature. Using FastAPI with sse-starlette + async integration tests with HTTPX testclient. Having a hard time testing my SSE endpoints due to lack of async testclients for SSE

**Comment by t94j0** [comment]
+1 This is the recommended option with FastAPI and sse-starlette. Would be really nice to have this feature

**Comment by harshitsinghai77** [comment]
Is this issue open? I would like to contribute if this is open. Thanks.

**Comment by johtso** [comment]
Guessing this is still valid..

**Comment by tomchristie** [comment]
We should probably bump up the durations on this stale bot. (Wasn't expecting that it'd repeatedly attempt closes after the first attempt) 😐

**Comment by abersheeran** [comment]
Surprised that this issue has been open for over a year. I just recently added functionality to rpc.py and wrote a code for parsing SSE. If anyone is interested in this, I'll be making a PR to httpx this week.
https://github.com/abersheeran/rpc.py/blob/cc26e983de9257a105dc8c5fab487cb3430dbe71/rpcpy/client.py#L203-L221

**Comment by tomchristie** [comment]
Ok. Probably a good plan would be to sketch out the API you'd propose, before submitting a pull request.

**Comment by tomchristie** [comment]
We should also strongly consider if we *really* want to have this as functionality in the core package or not.
Feasible that we should drop this ticket, but allow for a third party implementation.

For example, this'd make for a perfectly reasonable API, without us having to add additional functionality to `httpx` itself...

```python
from httpx_sse import iter_sse

with httpx.stream("GET", "https://www.example.org/sse") as response:
    for sse in iter_sse(response):
        print(sse.data, sse.event, sse.id, sse.retry)
```

**Comment by florimondmanca** [comment]
I worked on an `httpx-sse` package that I justed pushed here: https://github.com/florimondmanca/httpx-sse

The slight API difference from what @tomchristie suggested is the presence of an actual `SSETransport`, which mainly deals with some spec compliance for now. I've thought about making it handle reconnections, but that would be a lot more work and may not necessarily fit all use cases, as implementation needs may vary. So I've let that to a "how-to" guide in the project README.

```python
import httpx
from httpx_sse import SSETransport, iter_sse

with httpx.Client(transport=SSETransport()) as client:
    with httpx.stream("GET", "http://localhost:8000/sse") as response:
        for sse in iter_sse(response):
            print(sse.event, sse.data, sse.id, sse.retry)
```

This has not been released to PyPI yet. I'd be happy for folks to try that out first. :-)
